Realm.Shared/extensions/RealmResultsCollectionChanged.cs

```diff
@@ -0,0 +1,150 @@
+        /// </summary>
+        /// <param name="results">The <see cref="RealmResults{T}"/ > to observe for changes.</param>
+        /// <param name="errorCallback">An error callback that will be invoked if the observing thread raises an error.</param>
+        /// <param name="coalesceMultipleChangesIntoReset">
+        /// When a lot of items have been added or removed at once it is more efficient to raise <see cref="INotifyCollectionChanged.CollectionChanged" /> once
+        /// with <see cref="NotifyCollectionChangedAction.Reset" /> instead of multiple times for every single change. Pass <c>true</c> to opt-in to this behavior.
+        /// </param>
+        /// <returns>An <see cref="ObservableCollection{T}" />-like object useful for MVVM databinding.</returns>
+        /// <seealso cref="RealmResults{T}.SubscribeForNotifications(RealmResults{T}.NotificationCallback)"/>
+        public static INotifyCollectionChanged ToNotifyCollectionChanged<T>(this RealmResults<T> results, Action<Exception> errorCallback, bool coalesceMultipleChangesIntoReset) where T : RealmObject
+        {
+            return new ReadOnlyObservableCollection<T>(new Adapter<T>(results, errorCallback, coalesceMultipleChangesIntoReset));
+        }
+
+        sealed class Adapter<T> : ObservableCollection<T> where T : RealmObject
+        {
+            readonly RealmResults<T> _results;
+            readonly IDisposable _token;
+            readonly Action<Exception> _errorCallback;
+            readonly bool _coalesceMultipleChangesIntoReset;
+
+            private bool _suspendNotifications;
+
+            internal Adapter(RealmResults<T> results, Action<Exception> errorCallback, bool coalesceMultipleChangesIntoReset) : base(results)
+            {
+                _results = results;
+                _errorCallback = errorCallback;
+                _coalesceMultipleChangesIntoReset = coalesceMultipleChangesIntoReset;
+
+                _token = results.SubscribeForNotifications(OnChange);
+            }
+
+            ~Adapter()
+            {
+                _token.Dispose();
+            }
+
+            void OnChange(RealmResults<T> sender, RealmResults<T>.ChangeSet change, Exception error)
+            {
+                if (error != null)
+                {
+                    _errorCallback(error);
+                }
+                else if (change != null)
+                {
+                    _suspendNotifications = _coalesceMultipleChangesIntoReset && change.InsertedIndices.Length + change.DeletedIndices.Length > 1;
+
+                    foreach (var removed in change.DeletedIndices.Reverse())
+                    {
+                        // the row has been deleted, we need to recreate the adapter's contents
+                        if (!this[removed].RowHandle.IsAttached)
+                        {
+                            Recreate();
```

Comment:
Only if the current batch contains an object that has been deleted from the Realm. Such an object is invalid so we can't remove it from the observable collection because doing so would raise `CollectionChanged` with the `Remove` action and pass an invalid object to subscribers. So it's much simpler to just rebuild the observable collection and raise `CollectionChanged` with `Reset` instead.